sdk/core/src/http_client.rs

```diff
@@ -0,0 +1,126 @@
+use crate::errors::{AzureError, UnexpectedHTTPResult};
+use async_trait::async_trait;
+use http::{Request, Response, StatusCode};
+use hyper::{self, body, Body};
+use hyper_rustls::HttpsConnector;
+
+#[async_trait]
+pub trait HttpClient: Send + Sync + std::fmt::Debug {
+    async fn execute_request(
+        &self,
+        request: Request<&[u8]>,
+    ) -> Result<Response<Vec<u8>>, Box<dyn std::error::Error + Sync + Send>>;
+
+    async fn execute_request_check_status(
+        &self,
+        request: Request<&[u8]>,
+        expected_status: StatusCode,
+    ) -> Result<Response<Vec<u8>>, Box<dyn std::error::Error + Sync + Send>> {
+        let response = self.execute_request(request).await?;
+        if expected_status != response.status() {
+            Err(Box::new(AzureError::from(UnexpectedHTTPResult::new(
+                expected_status,
+                response.status(),
+                std::str::from_utf8(response.body())?,
+            ))))
+        } else {
+            Ok(response)
+        }
+    }
+
+    async fn execute_request_check_statuses(
+        &self,
+        request: Request<&[u8]>,
+        expected_statuses: &[StatusCode],
+    ) -> Result<Response<Vec<u8>>, Box<dyn std::error::Error + Sync + Send>> {
+        let response = self.execute_request(request).await?;
+        if !expected_statuses
+            .iter()
+            .any(|expected_status| *expected_status == response.status())
+        {
+            if expected_statuses.len() == 1 {
+                Err(Box::new(AzureError::from(UnexpectedHTTPResult::new(
+                    expected_statuses[0],
+                    response.status(),
+                    std::str::from_utf8(response.body())?,
+                ))))
+            } else {
+                Err(Box::new(AzureError::from(
+                    UnexpectedHTTPResult::new_multiple(
+                        expected_statuses.to_vec(),
+                        response.status(),
+                        std::str::from_utf8(response.body())?,
+                    ),
+                )))
+            }
+        } else {
+            Ok(response)
+        }
+    }
+}
+
+pub static EMPTY_BODY: [u8; 0] = [];
+
+#[async_trait]
+impl HttpClient for hyper::Client<HttpsConnector<hyper::client::HttpConnector>> {
+    async fn execute_request(
+        &self,
+        request: Request<&[u8]>,
+    ) -> Result<Response<Vec<u8>>, Box<dyn std::error::Error + Sync + Send>> {
+        let mut hyper_request = hyper::Request::builder()
+            .uri(request.uri())
+            .method(request.method());
+
+        for header in request.headers() {
+            hyper_request = hyper_request.header(header.0, header.1);
+        }
+
+        let body = String::from_utf8(request.body().to_vec())?;
```

Comment:
Yes you are right, I've added it to get an owned string but the simpler `request.body().to_vec()` will suffice. 
Ideally I would like to avoid allocating at all but `Body` requires a `'static` lifetime` :man_shrugging: